source/common/upstream/deterministic_aperture_lb.cc

```diff
@@ -0,0 +1,150 @@
+    return nullptr;
+  }
+
+  const std::pair<size_t, size_t> index_pair = pick2();
+
+  const RingHashLoadBalancer::RingEntry& first = ring_[index_pair.first];
+  const RingHashLoadBalancer::RingEntry& second = ring_[index_pair.second];
+
+  ENVOY_LOG(debug, "pick2 returned hosts: (hash1: {}, address1: {}, hash2: {}, address2: {})",
+            first.hash_, first.host_->address()->asString(), second.hash_,
+            second.host_->address()->asString());
+
+  return first.host_->stats().rq_active_.value() < second.host_->stats().rq_active_.value()
+             ? first.host_
+             : second.host_;
+}
+
+using HashFunction = envoy::config::cluster::v3::Cluster::RingHashLbConfig::HashFunction;
+DeterministicApertureLoadBalancer::Ring::Ring(
+    double offset, double width, const NormalizedHostWeightVector& normalized_host_weights,
+    double min_normalized_weight, uint64_t min_ring_size, uint64_t max_ring_size,
+    HashFunction hash_function, bool use_hostname_for_hashing, Stats::ScopeSharedPtr scope,
+    RingHashLoadBalancerStats ring_stats)
+    : RingHashLoadBalancer::Ring(normalized_host_weights, min_normalized_weight, min_ring_size,
+                                 max_ring_size, hash_function, use_hostname_for_hashing,
+                                 ring_stats),
+      offset_(offset), width_(width), unit_width_(1.0 / ring_size_), rng_(random_dev_()),
+      random_distribution_(0, 1), stats_(generateStats(*scope)) {
+  if (width_ > 1.0 || width_ < 0) {
+    throw EnvoyException(
+        fmt::format("Invalid width for the deterministic aperture ring{}", width_));
+  }
+}
+
+absl::optional<double> DeterministicApertureLoadBalancer::Ring::weight(size_t index, double offset,
+                                                                       double width) const {
+  if (index >= ring_size_ || width > 1 || offset > 1) {
+    return absl::nullopt;
+  }
+
+  double index_begin = index * unit_width_;
+  double index_end = index_begin + unit_width_;
+
+  if (offset + width > 1) {
+    double start = std::fmod((offset + width), 1.0);
+
+    return 1.0 - (intersect(index_begin, index_end, start, offset)) / unit_width_;
+  }
+
+  return intersect(index_begin, index_end, offset, offset + width) / unit_width_;
+}
+
+size_t DeterministicApertureLoadBalancer::Ring::getIndex(double offset) const {
+  RELEASE_ASSERT(offset >= 0 && offset <= 1, "valid offset");
+  return offset / unit_width_;
+}
+
+double DeterministicApertureLoadBalancer::Ring::intersect(double b0, double e0, double b1,
+                                                          double e1) const {
+  ENVOY_LOG(trace, "Overlap for (b0: {}, e0: {}, b1: {}, e1: {})", b0, e0, b1, e1);
+  return std::max(0.0, std::min(e0, e1) - std::max(b0, b1));
+}
+
+size_t DeterministicApertureLoadBalancer::Ring::pick() const {
+  return getIndex(std::fmod((offset_ + width_ * nextRandom()), 1.0));
+}
+
+size_t DeterministicApertureLoadBalancer::Ring::tryPickSecond(size_t first) const {
```

Comment:
Should this be `pickSecond` now that it always returns a pick?

Reply:
Good point @zuercher . I have updated the PR.